lib/timers.js

```diff
@@ -54,99 +112,134 @@ exports.active = function(item) {
-      var domain = first.domain;
-      if (domain && domain._disposed)
+      // https://github.com/nodejs/node-v0.x-archive/issues/2631
+      if (domain._disposed)
         continue;
 
-      try {
-        if (domain)
-          domain.enter();
-        threw = true;
-        first._called = true;
-        first._onTimeout();
-        if (domain)
-          domain.exit();
-        threw = false;
-      } finally {
-        if (threw) {
-          // We need to continue processing after domain error handling
-          // is complete, but not by using whatever domain was left over
-          // when the timeout threw its exception.
-          var oldDomain = process.domain;
-          process.domain = null;
-          process.nextTick(listOnTimeoutNT, list);
-          process.domain = oldDomain;
-        }
-      }
+      domain.enter();
     }
+
+    _tryOnTimeout(timer, list);
+
+    if (domain)
+      domain.exit();
   }
 
+  // If `L.peek(list)` returned nothing, the list was either empty or we have
+  // called all of the timer timeouts.
+  // As such, we can remove the list and clean up the TimerWrap C++ handle.
   debug('%d list empty', msecs);
   assert(L.isEmpty(list));
-  list.close();
-  delete lists[msecs];
+  this.close();
+  if (list._unrefed === true) {
+    delete unrefedLists[msecs];
+  } else {
+    delete refedLists[msecs];
+  }
+}
+
+
+// An optimization so that the try/finally only de-optimizes what is in this
+// smaller function.
+function _tryOnTimeout(timer, list) {
+  timer._called = true;
+  var threw = true;
+  try {
+    timer._onTimeout();
+    threw = false;
+  } finally {
+    if (!threw) return;
+
+    // We need to continue processing after domain error handling
+    // is complete, but not by using whatever domain was left over
+    // when the timeout threw its exception.
+    const domain = process.domain;
+    process.domain = null;
+    // If we threw, we need to process the rest of the list in nextTick.
+    process.nextTick(listOnTimeoutNT, list);
+    process.domain = domain;
+  }
 }
 
 
 function listOnTimeoutNT(list) {
-  list[kOnTimeout]();
+  list._timer[kOnTimeout]();
 }
 
 
+// A convenience function for re-using TimerWrap handles more easily.
+//
+// This mostly exists to fix https://github.com/nodejs/node/issues/1264.
+// Handles in libuv take at least one `uv_run` to be registered as unreferenced.
+// Re-using an existing handle allows us to skip that, so that a second `uv_run`
+// will return no active handles, even when running `setTimeout(fn).unref()`.
 function reuse(item) {
   L.remove(item);
 
-  var list = lists[item._idleTimeout];
+  var list = refedLists[item._idleTimeout];
   // if empty - reuse the watcher
   if (list && L.isEmpty(list)) {
     debug('reuse hit');
-    list.stop();
-    delete lists[item._idleTimeout];
-    return list;
+    list._timer.stop();
+    delete refedLists[item._idleTimeout];
+    return list._timer;
   }
```

Comment:
OK, I initially thought that the following sample code:

```
var timers = require('timers');

var internalTimerObject = {};
timers.enroll(internalTimerObject, 1000);

setTimeout(function timeoutA() {	
	timers.unenroll(internalTimerObject);

	setTimeout(function timeoutB() {

	}, 1000);
}, 1000);
```
would stop (not close, sorry for the misunderstanding) the underlying `TimerWrap` instance for "external" timers with a 1000ms delay and not restart it. 

After double checking, you're right that it doesn't: the "external" timer is still stopped when calling `unenroll` on the "internal" timer, but it's restarted again on the second `setTimeout` call. I think it shows a problem in the way the `unenroll` API cannot distinguish between "internal" and "external" timers. 

The problem itself was was not introduced by this PR but since this PR is about refactoring/unifying internal/external timers and making the implementation clearer/more consistent, I'm wondering when would be a good time to think about fixing that problem.